LibJS: for-await-of

davidot opened the pull request:
Since master has a broken Await AO this wont pass as many test as when #10933 is merged.

(Although that PR still isn't following the spec exactly I don't think we can actually do it as the spec says.)

davidot commented on Userland/Libraries/LibJS/Runtime/AsyncFromSyncIteratorPrototype.cpp:
```diff
@@ -0,0 +1,183 @@
+}
+
+// 27.2.1.1.1 IfAbruptRejectPromise ( value, capability ), https://tc39.es/ecma262/#sec-ifabruptrejectpromise
+#define TRY_OR_REJECT(vm, expression, capability)                                                         \
```
This macro is copied from PromiseConstructor.cpp
Should we maybe make this a macro you can import? Or somehow wrap this in a normal function call?
(I happen to know we will need this in at least one other spot)

linusg replied:
Obviously we don't want to have it duplicated. Belongs in `PromiseReaction.h` next to `27.2.1.1 PromiseCapability Records`.

linusg replied:
...and then include that header where it's needed.

davidot commented on Userland/Libraries/LibJS/Runtime/AsyncFromSyncIteratorPrototype.cpp:
```diff
@@ -0,0 +1,183 @@
+        if (_temporary_result.is_error()) {                                                               \
+            vm.clear_exception();                                                                         \
+            vm.stop_unwind();                                                                             \
+                                                                                                          \
+            /* a. Perform ? Call(capability.[[Reject]], undefined, « value.[[Value]] »). */             \
+            (void)vm.call(*capability.reject, js_undefined(), _temporary_result.release_error().value()); \
```
Sad clang-format noises but this is what it does

linusg replied:
I know, we alreadfy have this elsewhere. It's getting confused about the length of `«` and `»` I thikn, but as long as the formatting on CI matches it's ok.

davidot commented:
Fixed the issues, but unfortunately right now some test262 tests are flaky until #10933 is merged I think

linusg commented:
> Fixed the issues, but unfortunately right now some test262 tests are flaky until #10933 is merged I think

That PR is now merged, but I don't see how it would affect test262. LibJS doesn't use a `Core::EventLoop` itself.

linusg commented on Userland/Libraries/LibJS/Runtime/AsyncFromSyncIteratorPrototype.cpp:
```diff
@@ -0,0 +1,163 @@
+}
+
+// 27.1.4.4 AsyncFromSyncIteratorContinuation ( result, promiseCapability ), https://tc39.es/ecma262/#sec-asyncfromsynciteratorcontinuation
+static Object* async_from_sync_iterator_continuation(VM& vm, GlobalObject& global_object, Object& result, PromiseCapability& promise_capability)
```
I don't know why it is being ignored in the current version of `TRY_OR_REJECT`, but `IfAbruptRejectPromise` contains `? Call()`, so I think we need to fix the macro and then this must return `ThrowCompletionOr`.

davidot replied:
All the callers do so with `!` but I made it return a TCO and called with `MUST` for now.

linusg commented on Userland/Libraries/LibJS/Runtime/AsyncFromSyncIteratorPrototype.cpp:
```diff
@@ -0,0 +1,163 @@
+    auto* value_wrapper_promise = static_cast<Promise*>(value_wrapper);
+
+    value_wrapper_promise->perform_then(
+        NativeFunction::create(global_object, "", [done](VM& vm, GlobalObject& global_object1) -> ThrowCompletionOr<Value> {
+            return create_iterator_result_object(global_object1, vm.argument(0), done);
```
no need for the 1 suffix, you can shadow the outer variable name.

linusg commented on Userland/Libraries/LibJS/Runtime/AsyncFromSyncIteratorPrototype.cpp:
```diff
@@ -0,0 +1,163 @@
+// 27.1.4.4 AsyncFromSyncIteratorContinuation ( result, promiseCapability ), https://tc39.es/ecma262/#sec-asyncfromsynciteratorcontinuation
+static Object* async_from_sync_iterator_continuation(VM& vm, GlobalObject& global_object, Object& result, PromiseCapability& promise_capability)
+{
+    auto done = TRY_OR_REJECT(vm, promise_capability, iterator_complete(global_object, result));
+    auto value = TRY_OR_REJECT(vm, promise_capability, iterator_value(global_object, result));
+    auto value_wrapper = TRY_OR_REJECT(vm, promise_capability, promise_resolve(global_object, *global_object.promise_constructor(), value));
+
+    VERIFY(is<Promise>(value_wrapper));
+    auto* value_wrapper_promise = static_cast<Promise*>(value_wrapper);
+
+    value_wrapper_promise->perform_then(
+        NativeFunction::create(global_object, "", [done](VM& vm, GlobalObject& global_object1) -> ThrowCompletionOr<Value> {
+            return create_iterator_result_object(global_object1, vm.argument(0), done);
+        }),
+        js_undefined(), promise_capability);
+    return promise_capability.promise;
```
Please add spec comments to all new code where applicable.

linusg commented on Userland/Libraries/LibJS/Runtime/AsyncFromSyncIteratorPrototype.cpp:
```diff
@@ -0,0 +1,163 @@
+    auto* this_object = TRY(typed_this_object(global_object));
+    auto promise_capability = TRY(new_promise_capability(global_object, global_object.promise_constructor()));
+    auto& sync_iterator = this_object->sync_iterator_record();
+    auto* return_method = TRY_OR_REJECT(vm, promise_capability, sync_iterator.value_of().get_method(global_object, vm.names.return_));
```
Please explicitly wrap objects in a `Value()` call instead of using their `value_of()` method.

linusg commented on Userland/Libraries/LibJS/Runtime/AsyncFromSyncIteratorPrototype.cpp:
```diff
@@ -0,0 +1,163 @@
+        return promise_capability.promise;
+    }
+
+    ThrowCompletionOr<Value> result { Value(0) };
```
hmm, that's a bit annoying that we have to provide a value. can we use `js_undefined()` instead though?

linusg commented on Userland/Libraries/LibJS/Runtime/AsyncFromSyncIteratorPrototype.cpp:
```diff
@@ -0,0 +1,163 @@
+    else
+        result = call(global_object, return_method, &sync_iterator);
+
+    // Since result comes from two expressions we do IfAbruptRejectPromise explicitly here.
```
not sure what this means - `TRY_OR_REJECT` should work fine for this?

linusg commented on Userland/Libraries/LibJS/Runtime/AsyncFromSyncIteratorPrototype.cpp:
```diff
@@ -0,0 +1,163 @@
+    auto* this_object = TRY(typed_this_object(global_object));
+    auto promise_capability = TRY(new_promise_capability(global_object, global_object.promise_constructor()));
+    auto& sync_iterator = this_object->sync_iterator_record();
+    auto* throw_method = TRY_OR_REJECT(vm, promise_capability, sync_iterator.value_of().get_method(global_object, vm.names.throw_));
```
same here, `Value()` instead of `value_of()`

linusg commented on Userland/Libraries/LibJS/Runtime/IteratorOperations.cpp:
```diff
@@ -88,7 +96,9 @@ ThrowCompletionOr<Object*> iterator_step(GlobalObject& global_object, Object& it
-Completion iterator_close(Object& iterator, Completion completion)
+// 7.4.8 AsyncIteratorClose ( iteratorRecord, completion ), https://tc39.es/ecma262/#sec-asynciteratorclose
+// NOTE: These only differ in that async awaits the inner value after the call.
+static Completion iterator_close_(Object& iterator, Completion completion, IteratorHint is_async)
```
`is_async` is the name of a boolean, please call it something else (iterator_hint, for example)

linusg commented on Userland/Libraries/LibJS/Runtime/IteratorOperations.cpp:
```diff
@@ -114,10 +124,13 @@ Completion iterator_close(Object& iterator, Completion completion)
-        else
+        } else {
             inner_result = result_or_error.release_value();
+            if (is_async == IteratorHint::Async)
```
`// d. If innerResult.[[Type]] is normal, set innerResult to Await(innerResult.[[Value]]).`

davidot commented:
Added spec comments but in AST it's complicated so I hope the current comments are clear enough.
Please let me know if I made a mistake with the comments as this is the first real time I added them.

davidot commented on Userland/Libraries/LibJS/Runtime/AsyncFromSyncIteratorPrototype.cpp:
```diff
@@ -0,0 +1,205 @@
+
+    // 4. Let syncIteratorRecord be O.[[SyncIteratorRecord]].
+    auto& sync_iterator_record = this_object->sync_iterator_record();
+    // 5. If value is present, then
+    //     a. Let result be IteratorNext(syncIteratorRecord, value).
+    // 6. Else,
+    //     a. Let result be IteratorNext(syncIteratorRecord).
+    // 7. IfAbruptRejectPromise(result, promiseCapability).
+    auto* result = TRY_OR_REJECT(vm, promise_capability,
+        (vm.argument_count() > 0 ? iterator_next(sync_iterator_record, vm.argument(0))
+                                 : iterator_next(sync_iterator_record)));
```
I wrapped all this into a ternary. The code is a little bit further from the spec text but does remove the weird initialization of a throw completion and the extra if.

davidot commented:
Also this:

    Fixed the issues, but unfortunately right now some test262 tests are flaky until #10933 is merged I think

Was about the fact that the `Await` Abstract Operation was broken since it used a popped execution context but that should be fine now.

linusg commented on Userland/Libraries/LibJS/Runtime/IteratorOperations.cpp:
```diff
@@ -136,6 +152,18 @@ Completion iterator_close(Object& iterator, Completion completion)
+    return iterator_close_(iterator, move(completion), IteratorHint::Sync);
+}
+
+// 7.4.8 AsyncIteratorClose ( iteratorRecord, completion ), https://tc39.es/ecma262/#sec-asynciteratorclose
```
hmm, 7.4.8 already exists below - are the section numbers outdated?

davidot replied:
It seems so, I updated the wrong number in the same commit I added the new method. 
Or should these kind of things be done in a separate commit?

linusg replied:
Fine to do it in the same when you notice it while adding something new :)

linusg reviewed (commented):
Looks great, couple of more suggestions. 2nd commit should probably also mention that it's not a pure move but also contains the call() TRY fix.

davidot commented on Userland/Libraries/LibJS/Runtime/AsyncIteratorPrototype.cpp:
```diff
@@ -0,0 +1,33 @@
+    define_native_function(*vm.well_known_symbol_async_iterator(), symbol_async_iterator, 0, attr);
+}
+
+AsyncIteratorPrototype::~AsyncIteratorPrototype() = default;
```
Ah I was already confused but didn't look hard enough for the other cases whoops.

davidot commented:
Added a bit about adding TRY to the second commit, fixed all the suggestions and fixed that in one spot the spec text was not inside the if.
